src/ray/core_worker/reference_count.cc

```diff
@@ -103,20 +195,79 @@ void ReferenceCounter::DeleteReferences(const std::vector<ObjectID> &object_ids)
+    it->second.local_ref_count = 0;
+    it->second.submitted_task_ref_count = 0;
+    if (distributed_ref_counting_enabled_ && !it->second.CanDelete()) {
+      RAY_LOG(ERROR)
+          << "ray.internal.free does not currently work for objects that are still in "
+             "scope when distributed reference "
+             "counting is enabled. Try disabling ref counting by passing "
+             "distributed_ref_counting_enabled: 0 in the ray.init internal config.";
+    }
     DeleteReferenceInternal(it, nullptr);
   }
 }
 
-void ReferenceCounter::DeleteReferenceInternal(
-    absl::flat_hash_map<ObjectID, Reference>::iterator it,
-    std::vector<ObjectID> *deleted) {
-  if (it->second.on_delete) {
-    it->second.on_delete(it->first);
+void ReferenceCounter::DeleteReferenceInternal(ReferenceTable::iterator it,
+                                               std::vector<ObjectID> *deleted) {
+  const ObjectID id = it->first;
+  RAY_LOG(DEBUG) << "Attempting to delete object " << id;
+  if (distributed_ref_counting_enabled_ && it->second.RefCount() == 0 &&
+      it->second.on_local_ref_deleted) {
+    RAY_LOG(DEBUG) << "Calling on_local_ref_deleted for object " << id;
+    it->second.on_local_ref_deleted();
+    it->second.on_local_ref_deleted = nullptr;
   }
-  if (deleted) {
-    deleted->push_back(it->first);
+  PRINT_REF_COUNT(it);
+
+  bool evicted = false;
+  // If distributed ref counting is not enabled, then delete the object as soon
+  // as its local ref count goes to 0.
+  size_t local_ref_count =
+      it->second.local_ref_count + it->second.submitted_task_ref_count;
+  if (!distributed_ref_counting_enabled_ && local_ref_count == 0) {
+    RAY_LOG(DEBUG) << "Deleting object " << id;
+    if (it->second.on_delete) {
+      it->second.on_delete(id);
+      it->second.on_delete = nullptr;
+    }
+    if (deleted) {
+      deleted->push_back(id);
+    }
+    evicted = true;
+  }
+
+  if (it->second.CanDelete()) {
+    for (const auto &inner_id : it->second.contains) {
+      auto inner_it = object_id_refs_.find(inner_id);
+      if (inner_it != object_id_refs_.end()) {
+        RAY_LOG(DEBUG) << "Try to delete inner object " << inner_id;
+        if (it->second.owned_by_us) {
+          // If this object ID was nested in an owned object, make sure that
+          // the outer object counted towards the ref count for the inner
+          // object.
+          RAY_CHECK(inner_it->second.contained_in_owned.erase(id));
+        } else {
+          // If this object ID was nested in a borrowed object, make sure that
+          // we have already returned this information through a previous
+          // GetAndClearLocalBorrowers call.
+          RAY_CHECK(!inner_it->second.contained_in_borrowed_id.has_value());
+        }
+        DeleteReferenceInternal(inner_it, deleted);
+      }
+    }
+
+    if (!evicted) {
+      RAY_LOG(DEBUG) << "Deleting object " << id;
+      if (it->second.on_delete) {
+        it->second.on_delete(id);
```

Comment:
shouldn't we set `it->second.on_delete = nullptr` here like we do above?

Reply:
Oh I see, it's removed so we don't try to call it twice here. This chain of logic is pretty hard to follow (mostly due to the feature flag). Is it possible to just set a flag `should_delete=true` above in the local ref count check and then only do the deletion in one place at the bottom? Might make it clearer.